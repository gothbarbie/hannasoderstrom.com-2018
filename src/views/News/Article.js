// @flow
import React, { Component } from 'react'
import { Link, withRouter } from 'react-router-dom'
import { connect } from 'react-redux'

import styled from 'styled-components'

import { fetchRecently } from '../../actions'

import Wrapper from '../../components/Layout/Wrapper'
import LoadingSpinner from '../../components/LoadingSpinner'
import H1 from '../../components/Typography/H1'
import Small from '../../components/Typography/Small'

import type { RouterHistory } from 'react-router-dom'

const WrapperSmall = styled.section`
  max-width: 600px;
  margin: 5% auto;
  padding: 0 2rem;

  p {
    line-height: 150%;
    font-weight: 300;
    color: #4f4f4f;
    font-size: 18px;
    margin-bottom: 1rem;
  }

  img {
    width: 100%;
    height: auto;
  }
`

const Header = styled.header`
  margin-bottom: 2rem;
`

type Props = {
  loading: boolean,
  fetchRecently: Function,
  history: RouterHistory,
  recentlies: Array<{
    content: {
      rendered: string,
    },
    excerpt: {
      rendered: string,
    },
    date: string,
    title: {
      rendered: string,
    },
    media: {
      guid?: {
        rendered: string,
      },
    },
  }>,
}

export class Article extends Component<Props> {
  componentDidMount () {
    const url = decodeURI(window.location.pathname).split('/')
    const slug = url[2]
    const regex = new RegExp(/([A-Za-z0-9-])+/g)

    if (regex.test(decodeURI(slug))) {
      this.props.fetchRecently(slug)
    }
  }

  render () {
    if (this.props.loading) return <LoadingSpinner />

    if (!this.props.recentlies.length) return null

    const { content, date, title } = this.props.recentlies[0]

    const dateFormatted = new Date(date)
    const options = {
      year: 'numeric',
      month: 'long',
      day: 'numeric',
    }

    return (
      <Wrapper>
        <Link data-test="back-link" to="/news">
          ‚Üê Back
        </Link>
        <WrapperSmall>
          <Header>
            <H1
              dangerouslySetInnerHTML={{ __html: title.rendered }}
              data-test="article-title"
            />
            <Small
              dangerouslySetInnerHTML={{
                __html: dateFormatted.toLocaleDateString('en-EN', options),
              }}
            />
          </Header>
          <main
            className="main"
            dangerouslySetInnerHTML={{ __html: content.rendered }}
          />
        </WrapperSmall>
      </Wrapper>
    )
  }
}

const mapStateToProps = ({ recentlies, loading }) => {
  return {
    recentlies,
    loading,
  }
}

const mapDispatchToProps = {
  fetchRecently,
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(withRouter(Article))
