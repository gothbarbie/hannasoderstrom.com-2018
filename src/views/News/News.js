// @flow
import * as React from 'react'
import { Link } from 'react-router-dom'
import { connect } from 'react-redux'
import styled from 'styled-components'

import { fetchRecentlies } from '../../actions'

import Wrapper from '../../components/Layout/Wrapper'
import MainHeader from '../../components/Layout/MainHeader'
import MainFooter from '../../components/Layout/MainFooter'
import LoadingSpinner from '../../components/LoadingSpinner'
import Small from '../../components/Typography/Small'
import H1 from '../../components/Typography/H1'

import type { RecentlyType } from '../../types'

const Grid = styled.section`
  width: 100%;
  display: grid;
  grid-template-columns: 1fr;
  grid-template-rows: 1fr;
  grid-column-gap: 5%;
  grid-row-gap: 10%;
  align-items: start;

  .first {
    grid-column: 1 / span 1;
    grid-row: 1 / span 1;
  }

  @media (min-width: 700px) {
    grid-template-columns: 1fr 1fr 1fr;

    .first {
      grid-column: 1 / span 3;
    }
  }
`

const FillerWrapper = styled(Wrapper)`
  padding: 3rem 0;
`

const Article = styled.article`
  margin: 1rem 0;
  padding: 2rem 0;
  border-bottom: 1px solid #ee307f;

  h1,
  small {
    margin: 0 auto;
  }

  p {
    -webkit-font-smoothing: subpixel-antialiased;
    color: #333;
    font-size: 1rem;
    font-weight: 200;
    line-break: after-white-space;
    line-height: 1.8rem;
    max-width: 580px;
    overflow-wrap: break-word;
    padding: 1rem 0 1rem 0;
    width: 100%;
    word-wrap: break-word;
    margin: 0 auto;
  }

  .link-more {
    display: none;
  }

  img {
    width: 100%;
    height: auto;
    text-align: center;
    max-height: 180px;
    overflow: hidden;
  }

  &.first img {
    margin: 1.5rem auto;
    max-height: 100%;
  }

  &.first main {
    margin-top: 1.5vw;
  }
`

const Linked = styled(Link)`
  text-decoration: none;
  color: inherit;

  &:hover {
    text-decoration: underline;
  }
`

const SmallBlock = styled(Small)`
  display: block;
`

const ReadMoreLink = styled(Link)`
  display: block;
  margin-top: 2rem;
  text-align: right;
  color: #ee307f;
`

const H2 = styled.h2`
  font-size: 1.5rem;
  font-weight: 500;
  margin: 0;
`

type Props = {
  loading: boolean,
  fetchRecentlies: Function,
  recentlies: RecentlyType[],
}

export class News extends React.Component<Props> {
  componentDidMount () {
    this.props.fetchRecentlies()
  }

  renderArticles () {
    const { recentlies } = this.props

    return recentlies.map(item => {
      const dateFormatted = new Date(item.date)
      const options = {
        year: 'numeric',
        month: 'long',
        day: 'numeric',
      }

      const { id, content, slug, title, excerpt } = item

      const isFirst = recentlies.indexOf(item) === 0

      return (
        <Article className={isFirst && 'first'} data-test="article" key={id}>
          {!isFirst ? (
            <Linked to={`news/${slug}`}>
              <H2 dangerouslySetInnerHTML={{ __html: title.rendered }} />
            </Linked>
          ) : (
            <H1 dangerouslySetInnerHTML={{ __html: title.rendered }} />
          )}
          <SmallBlock
            dangerouslySetInnerHTML={{
              __html: dateFormatted.toLocaleDateString('en-EN', options),
            }}
          />
          {!isFirst &&
            item.media &&
            item.media.guid &&
            item.media.guid.rendered.length > 0 && (
              <img
                alt={title.rendered}
                src={item.media && item.media.guid.rendered}
              />
            )}
          {isFirst ? (
            <main dangerouslySetInnerHTML={{ __html: content.rendered }} />
          ) : (
            <main dangerouslySetInnerHTML={{ __html: excerpt.rendered }} />
          )}
          {!isFirst && (
            <ReadMoreLink to={`news/${slug}`}>Read more...</ReadMoreLink>
          )}
        </Article>
      )
    })
  }

  render () {
    return (
      <React.Fragment>
        <MainHeader darkColor />
        {this.props.recentlies.length ? (
          <Wrapper>
            <Grid>{this.renderArticles()}</Grid>
          </Wrapper>
        ) : (
          <React.Fragment>
            <FillerWrapper>
              <H2>Loading content...</H2>
            </FillerWrapper>
            <LoadingSpinner />
          </React.Fragment>
        )}
        <MainFooter />
      </React.Fragment>
    )
  }
}

const mapStateToProps = ({ recentlies, loading }) => {
  return {
    recentlies,
    loading,
  }
}

const mapDispatchToProps = {
  fetchRecentlies,
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(News)
