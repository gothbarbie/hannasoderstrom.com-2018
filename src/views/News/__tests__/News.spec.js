import React from 'react'
import { shallow } from 'enzyme'

import { News } from '../News'

describe('views/News', () => {
  let component
  let store
  let fetchRecentlies

  beforeEach(() => {
    fetchRecentlies = jest.fn()
    store = {
      dispatch: jest.fn(),
      getState: jest.fn(),
      subscribe: jest.fn(),
    }
    const recentlies = []
    component = shallow(
      <News
        fetchRecentlies={fetchRecentlies}
        recentlies={recentlies}
        store={store}
      />
    )
  })

  it('renders when no content', () => {
    expect(component).toMatchSnapshot()
  })

  it('renders with content', () => {
    component.setProps({
      recentlies: [
        {
          id: 1,
          title: { rendered: 'Hello' },
          slug: 'hello',
          excerpt: {
            rendered: '<p>Hello</p>',
          },
          date: 'today',
          content: { rendered: '<p>Hello world</p>' },
          media: {
            guid: {
              rendered: 'http://www.image.com/image01.jpg',
            },
          },
        },
      ],
    })
    expect(component).toMatchSnapshot()
  })

  describe('#fetchRecentlies', () => {
    it('has been called on mount', () => {
      expect(fetchRecentlies).toHaveBeenCalled()
    })
  })
})
