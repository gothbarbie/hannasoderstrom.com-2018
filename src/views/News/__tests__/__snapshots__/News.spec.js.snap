// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`views/News renders when no content 1`] = `
<React.Fragment>
  <MainHeader
    darkColor={true}
  />
  <React.Fragment>
    <Styled(styled.div)>
      <styled.h2>
        Loading content...
      </styled.h2>
    </Styled(styled.div)>
    <LoadingSpinner />
  </React.Fragment>
  <MainFooter />
</React.Fragment>
`;

exports[`views/News renders with content 1`] = `
<React.Fragment>
  <MainHeader
    darkColor={true}
  />
  <styled.div>
    <styled.section>
      <styled.article
        className="first"
        data-test="article"
        key="1"
      >
        <styled.h1
          dangerouslySetInnerHTML={
            Object {
              "__html": "Hello",
            }
          }
        />
        <Styled(styled.small)
          dangerouslySetInnerHTML={
            Object {
              "__html": "Invalid Date",
            }
          }
        />
        <main
          dangerouslySetInnerHTML={
            Object {
              "__html": "<p>Hello world</p>",
            }
          }
        />
      </styled.article>
    </styled.section>
  </styled.div>
  <MainFooter />
</React.Fragment>
`;
